src/options.c

```diff
@@ -0,0 +1,303 @@
+            return tcti_map_table[i].type;
+out:
+    return UNKNOWN_TCTI;
+}
+/*
+ * Convert from an element in the TCTI_TYPE enumeration to a string
+ * representation.
+ */
+char* const
+tcti_name_from_type (TCTI_TYPE tcti_type)
+{
+    int i;
+    for (i = 0; i < N_TCTI; ++i)
+        if (tcti_type == tcti_map_table[i].type)
+            return tcti_map_table[i].name;
+    return NULL;
+}
+/*
+ * Test a common_opts_t structure to be sure the member data has been
+ * populated. We don't do any tests on the data for appropriate formats
+ * (like testing socket_address for a valid IP address).
+ * return 0 if sanity test passes
+ * return 1 if help message was explicitly requested
+ * return 2 if sanity test fails
+ */
+int
+sanity_check_common (common_opts_t  *opts)
+{
+    if (opts->help)
+        return 1;
+    switch (opts->tcti_type) {
+#ifdef HAVE_TCTI_DEV
+    case DEVICE_TCTI:
+        if (opts->device_file == NULL) {
+            fprintf (stderr, "missing --device-file, see --help\n");
+            return 2;
+        }
+        break;
+#endif
+#ifdef HAVE_TCTI_SOCK
+    case SOCKET_TCTI:
+        if (opts->socket_address == NULL) {
+            fprintf (stderr, "missing --socket-address, see --help\n");
+            return 2;
+        }
+        if (opts->socket_port == 0) {
+            fprintf (stderr, "missing --socket-port, see --help\n");
+            return 2;
+        }
+        break;
+#endif
+    default:
+        fprintf (stderr, "invalid TCTI, see --help\n");
+        return 2;
+    }
+    return 0;
+}
+/*
+ * Populate the provided common_opts_t structure with data provided through
+ * the environment.
+ */
+void
+get_common_opts_from_env (common_opts_t *common_opts)
+{
+    char *env_str, *end_ptr;
+
+    if (common_opts == NULL)
+        return;
+    env_str = getenv (TPM2TOOLS_ENV_TCTI_NAME);
+    if (env_str != NULL)
+        common_opts->tcti_type = tcti_type_from_name (env_str);
+    env_str = getenv (TPM2TOOLS_ENV_DEVICE_FILE);
+    if (env_str != NULL)
+        common_opts->device_file = env_str;
+    env_str = getenv (TPM2TOOLS_ENV_SOCKET_ADDRESS);
+    if (env_str != NULL)
+        common_opts->socket_address = env_str;
+    env_str = getenv (TPM2TOOLS_ENV_SOCKET_PORT);
+    if (env_str != NULL)
+        common_opts->socket_port = strtol (env_str, &end_ptr, 10);
+}
+/*
+ * Get data from the environment and the caller (by way of the argument
+ * vector) to populate the provided common_opts_t structure. The data we get
+ * from the command line / argument vector takes presedence so we fill in the
+ * structure with data from the environment first, then from argv. Anything
+ * retrieved from the environment will just be over written with whatever we
+ * get from argv.
+ */
+int
+get_common_opts (int                     argc,
+                 char                   *argv[],
+                 common_opts_t          *common_opts)
+{
+    int c = 0, option_index = 0, ret = 0;
+    char *arg_str = "t:d:n:p:hvV";
```

Comment:
n: should be a typo for a: above. Further, I would suggest using '-T' for "--tcti", '-R' for "--socket-address" to avoid conflicts with short options used in existing tools. Change short options in existing tools will cause confusion for users.